!pip install openpyxl

# hc-3 documentation details

This notebook intends to give a closer look on the dataset's authors perspective on data organization.

If your intention is to load the documentation to memory you may skip it and go to the next notebook.

## Raw documentation

In [2]:
dataset_directory = 'data/hc-3'

In [3]:
import os

In [4]:
dataset_docs_directory = os.path.join(dataset_directory,'docs')

In [5]:
os.listdir(dataset_docs_directory)

['crcns-hc3-data-description.docx',
 '__MACOSX',
 '.ipynb_checkpoints',
 'hc3-original-docs',
 'crcns-hc3-data-description.pdf',
 'channelorder',
 'hc3-metadata-tables']

In this journey through hc-3 we shall pass through the calm waters of hc3-metadata-tables

### The hc3-metadata-tables

In [6]:
dataset_metadata_tables_directory = os.path.join(dataset_docs_directory,'hc3-metadata-tables')

#### The csv files

For the sake of simplicity let's load the CSV files employing the Pandas library

In [7]:
import pandas as pd

In [8]:
dataset_metadata_tables_csv_filenames = [f for f in os.listdir(dataset_metadata_tables_directory) if f.endswith('.csv')]

Let's spice things up with a little bit of regular expressions

In [9]:
import re

In [10]:
dataset_metadata_tables_csv_filekeys = [re.findall('\-([a-z]+)\.',filename)[0] for filename in dataset_metadata_tables_csv_filenames]

In [11]:
dataset_metadata_tables_csv = {}
for filekey,filename in zip(dataset_metadata_tables_csv_filekeys, dataset_metadata_tables_csv_filenames):
    filepath = os.path.join(dataset_metadata_tables_directory,filename)
    df = pd.read_csv(filepath , header=None)
    df = df.fillna(value='')
    dataset_metadata_tables_csv[filekey] = df 

In [12]:
print(dataset_metadata_tables_csv['session'].shape)
dataset_metadata_tables_csv['session'].sample(10).T

(442, 6)


158        237        312          56         329          48   \
0        465        685       1035          204       1078          180   
1   ec013.36   ec013.49   ec015.03   ec012ec.24   ec016.17   ec012ec.22   
2  ec013.610  ec013.880  ec015.046  ec012ec.496  ec016.233  ec012ec.455   
3     Mwheel     linear      sleep    bigSquare     linear    bigSquare   
4         10         10         10           10          7           10   
5   3592.602   1206.682      600.2       1154.0     1928.1        754.6   

         175        164                424        243  
0        515        488               1527        695  
1   ec013.38   ec013.37         g01_maze04   ec013.49  
2  ec013.669  ec013.639  g01_maze04_MS.006  ec013.890  
3     linear     linear             Mwheel  bigSquare  
4         10         10                 10         10  
5      798.3   1400.422            1125.99   1291.469

In [13]:
print(dataset_metadata_tables_csv['files'].shape)
dataset_metadata_tables_csv['files'].sample(10).T

(442, 5)


304         345        196          24         166         337  \
0    ec014.42    ec016.37   ec013.41   ec012ec.16   ec013.37    ec016.27   
1   ec014.797   ec016.582  ec013.743  ec012ec.312  ec013.643   ec016.430   
2  1717294379  2732801593  611665141    338633580  321765649  3667720900   
3         mpg           -          -          m1v          -           -   
4   485707780           0          0    101328345          0           0   

         73           28                  402                 315  
0   ec013.18   ec012ec.17          gor01-6-12          ec014.n329  
1  ec013.205  ec012ec.350  2006-6-12_15-55-31  2007-3-29_16-06-46  
2   55571161    359398649          1194376673          7672546866  
3          -          m1v                   -                   -  
4          0    104065218                   0                   0

In [14]:
print(dataset_metadata_tables_csv['cell'].shape)
dataset_metadata_tables_csv['cell'].sample(10).T

(7736, 15)


1384        7427      1747      669       2714       4943      4660  \
0        1455        7896      1844       703      2841       5119      4835   
1    ec013.36  i01_maze05  ec013.40  ec013.28  ec013.50   ec016.30  ec016.19   
2       ec013       i01_m     ec013     ec013     ec013      ec016     ec016   
3           6           9         6         4         8          5        11   
4          11          15         6         6         2          4         2   
5         CA1         CA1       CA1       EC5        DG        EC3       EC5   
6           0           0         1         0         1          0         1   
7          11           0         0         0         0          3         0   
8           0           0         1         0         1          0         1   
9           1           0         0         0         0          1         0   
10         17           0         0         0         0          0         0   
11          2           0         1         0         0          1         1   
12   38.44919    0.152884  0.758082   0.37993  0.119397  14.823418  1.527165   
13  38.444614    0.149604  0.757753  0.385046  0.165837  17.725886  2.601309   
14          i           p         p         p         p          i         p   

        1778      3925      2565  
0       1879      4087      2689  
1   ec013.40  ec014.25  ec013.49  
2      ec013     ec014     ec013  
3          8         4         1  
4         13        10         6  
5        CA1       CA1       EC2  
6          0         1         0  
7          0         0         0  
8          0         1         0  
9          0         0         0  
10         0         0         0  
11         0         2         0  
12  2.434443  2.281335  0.506069  
13  4.768342  2.382157  0.861401  
14         p         p         p

In [15]:
print(dataset_metadata_tables_csv['epos'].shape)
dataset_metadata_tables_csv['epos'].sample(10).T

(119, 18)


84        79        53        54        73          98        60  \
0   ec016.45  ec016.37  ec014.16  ec014.17  ec016.25  f01_maze08  ec014.26   
1      ec016     ec016     ec014     ec014     ec016       f01_m     ec014   
2         DG       CA1       CA1       CA1       CA1         CA1       CA1   
3         DG       CA1       CA1       CA1       CA1         CA1       CA1   
4         DG       CA1       CA1       CA1       CA1         CA1       CA1   
5        CA3       CA1       CA1       CA1       CA1         CA1       CA1   
6        EC2       EC3       CA1       CA1       EC3         CA1       CA1   
7        EC2       EC4       CA1       CA1       EC4         CA1       CA1   
8        CA3       CA1       CA1       CA1       CA1         CA1       CA1   
9        CA3       CA1       CA1       CA1       CA1         CA1       CA1   
10       CA3       CA1       EC2       EC2       CA1                   EC2   
11       CA3       CA1       EC3       EC3       CA1                   EC3   
12       EC3       EC5       EC3       EC3       EC5                   EC3   
13       EC5       EC5       EC5       EC5       EC5                   EC5   
14                                                                           
15                                                                           
16                                                                           
17                                                                           

          24        89        90  
0   ec013.30  ec016.50  ec016.51  
1      ec013     ec016     ec016  
2        EC3        DG        DG  
3        EC4        DG        DG  
4        EC5        DG        DG  
5        EC5       CA3       CA3  
6        CA1       EC2       EC2  
7        CA1       EC2       EC2  
8        CA1       CA3       CA3  
9        CA1       CA3       CA3  
10                 CA3       CA3  
11                 CA3       CA3  
12                 EC3            
13                 EC5       EC3  
14                           EC5  
15                                
16                                
17

#### The xlsx file

In [16]:
dataset_metadata_tables_xlsx_file = os.path.join(dataset_metadata_tables_directory,'hc3-tables.xlsx')

In [17]:
import openpyxl

In [18]:
dataset_metadata_tables_xlsx_workbook = openpyxl.load_workbook(dataset_metadata_tables_xlsx_file)

/home/jose/miniconda3/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


In [19]:
dataset_metadata_tables_xlsx_workbook.sheetnames

['cell', 'session', 'file', 'epos']

In [20]:
dataset_metadata_tables_xlsx = {}
for sheetname in dataset_metadata_tables_xlsx_workbook.sheetnames:
    # a bit of a hack to get the header row in the right place
    # https://stackoverflow.com/a/51786976/454773
    df = pd.read_excel(dataset_metadata_tables_xlsx_file, sheet_name=sheetname, header=3) 
    
    df.drop(df.columns[df.columns.str.contains('Unnamed')], axis=1, inplace=True)
    
    dataset_metadata_tables_xlsx[sheetname] = df.fillna(value='')

/home/jose/miniconda3/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)
/home/jose/miniconda3/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)
/home/jose/miniconda3/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


In [21]:
print(dataset_metadata_tables_xlsx['session'].shape)
dataset_metadata_tables_xlsx['session'].sample(10).T

(442, 6)


0                  423           398          55   \
id                  5               1525          1480          200   
topdir     ec012ec.11         g01_maze04   pin01-11-04   ec012ec.23   
session   ec012ec.187  g01_maze04_MS.004  11-05_0-24-7  ec012ec.491   
behavior       Mwheel             Mwheel         Tmaze       Mwheel   
fam                10                 10            10           10   
duration       1096.4           1314.406      566.4848       1068.0   

                288          24         122        229        307          37   
id              902           91        384        657       1030          140  
topdir     ec014.27   ec012ec.16   ec013.31   ec013.47   ec015.03   ec012ec.19  
session   ec014.405  ec012ec.312  ec013.504  ec013.846  ec015.041  ec012ec.398  
behavior     Mwheel       Mwheel     linear      wheel  midSquare       Mwheel  
fam              10           10         10         10          7           10  
duration   5054.874       1060.7      668.1   1521.254     2825.7       1137.8

In [22]:
print(dataset_metadata_tables_xlsx['file'].shape)
dataset_metadata_tables_xlsx['file'].sample(10).T

(442, 5)


429        98         193        91         140  \
topdir              vvp01-4-10   ec013.29   ec013.41   ec013.28   ec013.33   
session     2006-4-10_14-22-10  ec013.434  ec013.737  ec013.406  ec013.556   
size                1050482263  378004695  505338954  534875736  257003664   
vtype                        -        mpg          -          -          -   
video_size                   0  218073092          0          0          0   

                  84         205        125         337                 432  
topdir       ec013.27   ec013.42   ec013.32    ec016.27          vvp01-4-10  
session     ec013.386  ec013.765  ec013.525   ec016.430  2006-4-10_21-32-13  
size        217393634  448605822  610200232  3667720900          1195793952  
vtype               -        mpg          -           -                   -  
video_size          0  269889540          0           0                   0

In [23]:
print(dataset_metadata_tables_xlsx['epos'].shape)
dataset_metadata_tables_xlsx['epos'].sample(10).T

(119, 18)


65          0         87          6         58        55   \
topdir  ec014.39  ec012ec.11  ec016.48  ec012ec.17  ec014.24  ec014.18   
animal     ec014       ec012     ec016       ec012     ec014     ec014   
e1           CA1         EC3        DG         EC3       CA1       CA1   
e2           CA1         EC3        DG         EC3       CA1       CA1   
e3           CA1         EC5        DG         EC5       CA1       CA1   
e4           CA1         EC5       CA3         EC5       CA1       CA1   
e5           CA1                   EC2                   CA1       CA1   
e6           CA1                   EC2                   CA1       CA1   
e7           CA1                   CA3                   CA1       CA1   
e8           CA1                   CA3                   CA1       CA1   
e9           EC2                   CA3                   EC2       EC2   
e10          EC3                   CA3                   EC3       EC3   
e11          EC3                   EC3                   EC3       EC3   
e12          EC5                   EC5                   EC5       EC5   
e13                                                                      
e14                                                                      
e15                                                                      
e16                                                                      

             53        28          8           117  
topdir  ec014.16  ec013.34  ec012ec.19  vvp01-4-18  
animal     ec014     ec013       ec012       vvp01  
e1           CA1       EC2         EC3         CA3  
e2           CA1       EC3         EC3         CA3  
e3           CA1       EC4         EC5         CA3  
e4           CA1       EC5         EC5         CA3  
e5           CA1       CA1                     CA1  
e6           CA1       CA1                     CA1  
e7           CA1       CA1                     CA1  
e8           CA1       CA1                     CA1  
e9           EC2                                    
e10          EC3                                    
e11          EC3                                    
e12          EC5                                    
e13                                                 
e14                                                 
e15                                                 
e16

In [24]:
print(dataset_metadata_tables_xlsx['cell'].shape)
dataset_metadata_tables_xlsx['cell'].sample(10).T

(7736, 15)


964         7333        7537       6411        7571  \
id                  1019        7798        8006       6602        8040   
topdir          ec013.31  i01_maze04  i01_maze15  vvp01-4-9  i01_maze15   
animal             ec013       i01_m       i01_m      vvp01       i01_m   
ele                    7           9           7          4           8   
clu                    9          14           2          2          14   
region               CA1         CA1         CA1        CA1         CA1   
ne                     0           0           1          0           0   
ni                     0           0           0          0           0   
eg                     0           0           1          0           0   
ig                     0           0           0          0           0   
ed                     2           0           0          0           0   
id.1                   1           0           0          0           0   
fireRate        6.774116    0.137599    0.122465   0.251635    0.790573   
totalFireRate  13.084035    0.154732    0.132775   0.250856    0.787712   
type                   i           p           p          p           p   

                   3028        6352       5374      1247      3507  
id                 3157        6543       5554      1315      3663  
topdir         ec013.55  vvp01-4-18   ec016.44  ec013.35  ec014.17  
animal            ec013       vvp01      ec016     ec013     ec014  
ele                   1           6          4         2         6  
clu                  17           6         11         2         3  
region              EC2         CA1        CA3       EC3       CA1  
ne                    0           1          0         1         1  
ni                    0           0          1         0         0  
eg                    0           1          0         1         1  
ig                    0           0          1         0         0  
ed                    1           0          0         0         0  
id.1                  1           0          1         1         0  
fireRate       1.128031    0.883053  14.382088  0.219211  0.120487  
totalFireRate  1.225361    0.958728  14.391903  0.279484  0.117931  
type                  i           p          i         p         p

#### The db file

In [25]:
dataset_metadata_tables_db_filepath = os.path.join(dataset_metadata_tables_directory,'hc3-tables.db')

In [26]:
import sqlite3

In [27]:
import pandas as pd

In [28]:
dataset_metadata_tables_db_connection = sqlite3.connect(dataset_metadata_tables_db_filepath)
dataset_metadata_tables_db_dataframe = pd.read_sql_query("SELECT * FROM sqlite_master", dataset_metadata_tables_db_connection)
dataset_metadata_tables_db = {}

for table_name in dataset_metadata_tables_db_dataframe['tbl_name']:
    df = pd.read_sql_query('SELECT * from %s' % table_name, dataset_metadata_tables_db_connection)
    df = df.fillna(value='')
    df = df.replace('""','')
    df = df.replace('None','')
    
    dataset_metadata_tables_db[table_name] = df

dataset_metadata_tables_db_connection.close()

In [29]:
dataset_metadata_tables_db

{'session':        id      topdir            session   behavior  familiarity  duration
 0       5  ec012ec.11        ec012ec.187     Mwheel           10  1096.400
 1       6  ec012ec.11        ec012ec.188     Mwheel           10   171.600
 2       7  ec012ec.11        ec012ec.189     Mwheel           10   942.500
 3      19  ec012ec.12        ec012ec.209     Mwheel           10   220.300
 4      20  ec012ec.12        ec012ec.210     Mwheel           10   747.800
 ..    ...         ...                ...        ...          ...       ...
 437  1548  i01_maze15  i01_maze15_MS.002     Mwheel           10  1440.563
 438  1551  j01_maze04   j01_maze04_M.001  bigSquare           10  3802.300
 439  1553  j01_maze04   j01_maze04_M.003  bigSquare           10   984.100
 440  1554  j01_maze04   j01_maze04_M.004  bigSquare           10  2240.400
 441  1555  j01_maze05   j01_maze05_M.001  bigSquare           10  3570.500
 
 [442 rows x 6 columns],
 'epos':           topdir animal   e1   e2   e3   

In [30]:
print(dataset_metadata_tables_db['session'].shape)
dataset_metadata_tables_db['session'].sample(10).T

(442, 6)


159          19         289        213        336  \
id                 471           74        925        624       1119   
topdir        ec013.36   ec012ec.15   ec014.28   ec013.45   ec016.25   
session      ec013.620  ec012ec.285  ec014.440  ec013.805  ec016.405   
behavior     bigSquare       Mwheel     Mwheel     linear      wheel   
familiarity         10           10         10         10          6   
duration      1779.712       1115.2    4740.71   1227.162     2533.7   

                           425        162        121        171  \
id                        1529        477        383        504   
topdir              g01_maze05   ec013.36   ec013.31   ec013.38   
session      g01_maze05_MS.001  ec013.627  ec013.503  ec013.658   
behavior                Mwheel     linear     linear     Mwheel   
familiarity                 10         10         10         10   
duration                2074.0   1234.944      798.0     2644.2   

                            408  
id                         1494  
topdir               vvp01-4-18  
session      2006-4-18_13-06-01  
behavior              linearOne  
familiarity                  10  
duration               825.4417

In [31]:
print(dataset_metadata_tables_db['file'].shape)
dataset_metadata_tables_db['file'].sample(10).T

(442, 5)


400        120        118          9          265  \
topdir             g01_maze11   ec013.31   ec013.31   ec012ec.13   ec013.54   
session     g01_maze11_MS.002  ec013.502  ec013.492  ec012ec.228  ec013.955   
size               1530694606  261567812  196949373    330148992  368892764   
video_type                  -          -        mpg          m1v        mpg   
video_size                  0          0  138326020     98284109  208476164   

                  273                 404         347        202        146  
topdir       ec013.55          gor01-6-12    ec016.41   ec013.42   ec013.34  
session     ec013.974  2006-6-12_19-26-43   ec016.665  ec013.761  ec013.580  
size        587925508          2483580426  1519396699  401350515  452922272  
video_type        mpg                   -         mpg          -        mpg  
video_size  326096900                   0   779687940          0  168691716

In [32]:
print(dataset_metadata_tables_db['epos'].shape)
dataset_metadata_tables_db['epos'].sample(13).T

(119, 18)


107       70        47        84        96          101  \
topdir  gor01-6-7  ec016.11  ec013.53  ec016.45  ec016.58  f01_maze12   
animal      gor01     ec016     ec013     ec016     ec016       f01_m   
e1            CA3       CA1       EC2        DG        DG         CA1   
e2            CA3       CA1       EC3        DG        DG         CA1   
e3            CA3       CA1       EC4        DG        DG         CA1   
e4            CA3       CA1       EC5       CA3       CA3         CA1   
e5            CA3       EC3       CA3       EC2       EC2         CA1   
e6            CA3       EC4       CA3       EC2       EC2         CA1   
e7            CA3       CA1       CA3       CA3       CA3         CA1   
e8            CA3       CA1       CA3       CA3       CA3         CA1   
e9            CA1       CA1                 CA3       CA3               
e10           CA1       CA1                 CA3       CA3               
e11           CA1       EC5                 EC3       EC3               
e12           CA1       EC5                 EC5       EC5               
e13                                                                     
e14                                                                     
e15                                                                     
e16                                                                     

               13          110       43        23        69        32   \
topdir  ec012ec.24  i01_maze04  ec013.49  ec013.29  ec015.03  ec013.38   
animal       ec012       i01_m     ec013     ec013     ec014     ec013   
e1             EC3        EC11       EC2       EC3       CA1       EC2   
e2             EC3        EC12       EC3       EC4       CA1       EC3   
e3             EC5        EC13       EC4       EC5       CA1       EC4   
e4             EC5        EC14       EC5       EC5       CA1       EC5   
e5                         CA1       CA1       CA1       CA1       CA1   
e6                         CA1       CA1       CA1       CA1       CA1   
e7                         CA1       CA1       CA1       CA1       CA1   
e8                         CA1       CA1       CA1       CA1       CA1   
e9                        CA1c                           EC2             
e10                       CA1c                           EC3             
e11                       CA1c                           EC3             
e12                       CA1c                           EC5             
e13                       CA1c                                           
e14                                                                      
e15                                                                      
e16                                                                      

             51   
topdir  ec014.12  
animal     ec014  
e1           CA1  
e2           CA1  
e3           CA1  
e4           CA1  
e5           CA1  
e6           CA1  
e7           CA1  
e8           CA1  
e9           EC2  
e10          EC3  
e11          EC3  
e12          EC5  
e13               
e14               
e15               
e16

In [33]:
print(dataset_metadata_tables_db['cell'].shape)
dataset_metadata_tables_db['cell'].sample(10).T

(7736, 15)


7551      4767      4927      1273        7131      4080  \
id                   8020      4942      5103      1342        7590      4246   
topdir         i01_maze15  ec016.27  ec016.30  ec013.35  g01_maze05  ec014.28   
animal              i01_m     ec016     ec016     ec013       g01_m     ec014   
ele                     7         3         3         5          14        11   
clu                    16         3         2         5           6         4   
region                CA1       CA1       CA1       CA1         CA1       EC3   
nexciting               0         0         0         0           1         1   
ninhibiting             0         0         0         0           0         0   
exciting                0         0         0         0           1         1   
inhibiting              0         0         0         0           0         0   
excited                 0         0         0         0           0         0   
inhibited               0         0         0         0           0         2   
fireRate         0.134457  0.084194  0.102369  0.430967    1.943109  3.449885   
totalFireRate    0.134055  0.084265  0.102177  0.780229    1.943272  3.452209   
cellType                p         p         n         p           p         p   

                     7451        140       5714      6036  
id                   7920         142      5899      6225  
topdir         i01_maze05  ec012ec.17  ec016.51  ec016.58  
animal              i01_m       ec012     ec016     ec016  
ele                    10           1         9         3  
clu                    11           3         3        10  
region                CA1         EC3       CA3        DG  
nexciting               0           0         0         0  
ninhibiting             0           0         0         0  
exciting                0           0         0         0  
inhibiting              0           0         0         0  
excited                 1           0         0         0  
inhibited               0           0         0         0  
fireRate         0.224103    1.637883  1.027133  0.541863  
totalFireRate    0.225198    1.798534  1.026374  0.541827  
cellType                p           i         p         p

#### Files comparison, a sanity check

We find that the different files yields the similar keys

In [34]:
sorted(dataset_metadata_tables_csv.keys())

['cell', 'epos', 'files', 'session']

In [35]:
sorted(dataset_metadata_tables_db.keys())

['cell', 'epos', 'file', 'session']

In [36]:
sorted(dataset_metadata_tables_xlsx.keys())

['cell', 'epos', 'file', 'session']

but not quite there yet...

In [37]:
dataset_metadata_tables_csv['file'] = dataset_metadata_tables_csv['files']

The shapes are the same

In [38]:
for key in ['cell', 'epos', 'file', 'session']:
    a = dataset_metadata_tables_csv[key]
    b = dataset_metadata_tables_xlsx[key]
    c = dataset_metadata_tables_db[key]
    
    print(a.shape == b.shape == c.shape, a.shape, b.shape, c.shape)


True (7736, 15) (7736, 15) (7736, 15)
True (119, 18) (119, 18) (119, 18)
True (442, 5) (442, 5) (442, 5)
True (442, 6) (442, 6) (442, 6)


In [39]:
for key in ['cell', 'epos', 'file', 'session']:
    a = dataset_metadata_tables_csv[key]
    b = dataset_metadata_tables_xlsx[key]
    c = dataset_metadata_tables_db[key]
    print((a.values == b.values).all(), (a.values == c.values).all(), (b.values == c.values).all(), key)
    

False False False cell
True True True epos
True True True file
True True True session


In [40]:
key = 'cell'

a = dataset_metadata_tables_csv[key]
b = dataset_metadata_tables_xlsx[key]
c = dataset_metadata_tables_db[key]

import numpy as np


id_ab_row = np.sum(a.values == b.values, axis=0) != a.shape[0]
id_ab_col = np.sum(a.values == b.values, axis=1) != a.shape[1]

id_ac_row = np.sum(a.values == c.values, axis=0) != a.shape[0]
id_ac_col = np.sum(a.values == c.values, axis=1) != a.shape[1]

id_bc_row = np.sum(b.values == c.values, axis=0) != b.shape[0]
id_bc_col = np.sum(b.values == c.values, axis=1) != b.shape[1]

print(np.allclose(a.iloc[id_ab_col,id_ab_row].replace('',-1),b.iloc[id_ab_col,id_ab_row].replace('',-1)))
print(np.allclose(a.iloc[id_ac_col,id_ac_row].replace('',-1),c.iloc[id_ac_col,id_ac_row].replace('NaN',-1)))
print(np.allclose(b.iloc[id_bc_col,id_bc_row].replace('',-1),c.iloc[id_bc_col,id_bc_row].replace('NaN',-1)))


True
True
True


The cell tables are different one to another yet close to a precision of relative ($10^{-5}$) and absolute ($10^{-8}$) tolerances.

I assume that I liked the sqlite database file more friendly

### The hc3-original-docs directory

In [41]:
dataset_originaldocs_directory = os.path.join(dataset_docs_directory,'hc3-original-docs')

In [42]:
os.listdir(dataset_originaldocs_directory)

['hc3.db',
 'README.txt',
 'DecidePyrIntRegion.m',
 '.ipynb_checkpoints',
 'KenjiData.mat',
 'tables.sql',
 'ElePosition.txt']

#### The hc3.db file

In [43]:
dataset_originaldocs_db_filepath = os.path.join(dataset_originaldocs_directory,'hc3.db')
dataset_originaldocs_db_connection = sqlite3.connect(dataset_originaldocs_db_filepath)
dataset_originaldocs_db_dataframe = pd.read_sql_query("SELECT * FROM sqlite_master", dataset_originaldocs_db_connection)
dataset_originaldocs_db = {}

for table_name in dataset_originaldocs_db_dataframe['tbl_name']:
    df = pd.read_sql_query('SELECT * from %s' % table_name, dataset_originaldocs_db_connection)
    df = df.fillna(value='')
    df = df.replace('""','')
    df = df.replace('None','')
    
    dataset_originaldocs_db[table_name] = df

dataset_originaldocs_db_connection.close()

In [44]:
{k:dataset_originaldocs_db[k].shape for k in dataset_originaldocs_db.keys()}

{'elepos': (133, 20),
 'pmMap': (8569, 44),
 'region_pm1943': (8569, 2),
 'region_elepos': (8569, 2),
 'beh': (1692, 9),
 'hc3_mpg': (125, 3),
 'clean': (8569, 2),
 'to_remove': (15, 3),
 'animal': (12, 2),
 'region': (8569, 2),
 'region_name': (9, 2),
 'map_cell_counts': (120, 2),
 'clu_cell_counts': (120, 3),
 'video_files': (198, 4),
 'hc3_session': (442, 4),
 'hc3_topdir': (119, 2),
 'iCell': (8569, 6)}

In [45]:
dataset_originaldocs_db['animal'].sample(10).T


3      2      0      1      11     8      5     6      4      9 
id        4      3      1      2     12      9      6     7      5     10
name  ec016  ec014  ec012  ec013  j01_m  vvp01  i01_m  km01  gor01  f01_m

In [46]:
dataset_originaldocs_db['hc3_topdir'].sample(10).T


102       64        52        34          0         59   \
topdir  g01_maze04  ec014.36  ec014.13  ec013.40  ec012ec.11  ec014.25   
animal       g01_m     ec014     ec014     ec013       ec012     ec014   

             60          104       82        27   
topdir  ec014.26  g01_maze11  ec016.43  ec013.33  
animal     ec014       g01_m     ec016     ec013

In [47]:
print(dataset_originaldocs_db['hc3_mpg'].shape)
dataset_originaldocs_db['hc3_mpg'].sample(10).T

(125, 3)


103        87         46         75         94         11   \
id             104         88         47         76         95         12   
topdir    ec016.33   ec014.42   ec013.44   ec014.26   ec016.11   ec013.30   
session  ec016.537  ec014.797  ec013.782  ec014.379  ec016.110  ec013.470   

               15         63         56         104  
id              16         64         57        105  
topdir    ec013.31   ec013.53   ec013.48   ec016.37  
session  ec013.491  ec013.938  ec013.856  ec016.590

In [48]:
print(dataset_originaldocs_db['hc3_session'].shape)
dataset_originaldocs_db['hc3_session'].sample(10).T

(442, 4)


35         141        380         375          56   \
id                36        142        381         376           57   
topdir    ec012ec.18   ec013.34   ec016.57    ec016.56   ec012ec.24   
session  ec012ec.375  ec013.570  ec016.985   ec016.957  ec012ec.498   
size       165519110  552046498  240078661  1952962505    326316024   

                        308                 312        196        334  \
id                      309                 313        197        335   
topdir           ec014.n329          ec014.n329   ec013.41   ec016.19   
session  2007-3-29_12-41-10  2007-3-29_14-32-45  ec013.743  ec016.275   
size              161678377           783039792  611665141  283947464   

               214  
id             215  
topdir    ec013.45  
session  ec013.806  
size     437093954

In [49]:
dataset_originaldocs_db['beh'].sample(10).T

1419          1598            1681           1335  \
id                      1420          1599            1682           1336   
topdir              ec016.58       km01.05        KM011206       ec016.51   
mergen       ec016.1002_1023  km01.041_053  KM011206_01_11  ec016.871_889   
unit                   unit4         unit5           unit5          unit5   
session           ec016.1002      km01.053     KM011206_01      ec016.882   
behavior               sleep         sleep           sleep          sleep   
familiarity               10            10              10             10   
duration               990.1         331.2           600.0          916.7   
problem                                                                     

                      982            1365           988              16    \
id                     983           1366            989               17   
topdir            ec014.39       ec016.53       ec014.41       ec012ec.12   
mergen       ec014.716_742  ec016.914_932  ec014.759_779  ec012ec.199-219   
unit                 unit5          unit6          unit4            unit6   
session          ec014.742      ec016.922      ec014.764      ec012ec.206   
behavior             sleep          sleep          sleep            sleep   
familiarity             10             10             10               10   
duration          881.0496          792.9        2549.76            762.2   
problem                                                                     

                      653           1605  
id                     654          1606  
topdir            ec013.47       km01.06  
mergen       ec013.839_851  km01.055_071  
unit                 unit4         unit5  
session          ec013.843      km01.066  
behavior            linear        circle  
familiarity             10             8  
duration          1599.488         427.1  
problem

In [50]:
dataset_originaldocs_db['video_files'].sample(10).T


91               3               196            181  \
topdir        ec013.33       ec012ec.12        ec016.58       ec016.46   
session      ec013.543      ec012ec.209      ec016.1016      ec016.775   
file     ec013.543.mpg  ec012ec.209.m1v  ec016.1016.mpg  ec016.775.mpg   
size          23717892         19923915       640278532      570927108   

                     2                25               0              97   \
topdir        ec012ec.11       ec012ec.17       ec012ec.11       ec013.34   
session      ec012ec.189      ec012ec.348      ec012ec.187      ec013.582   
file     ec012ec.189.m1v  ec012ec.348.m1v  ec012ec.187.m1v  ec013.582.mpg   
size            90686244        103308841        106726972       89505796   

                     60               56   
topdir        ec012ec.25       ec012ec.24  
session      ec012ec.524      ec012ec.498  
file     ec012ec.524.m1v  ec012ec.498.m1v  
size            97859850        102828296

In [51]:
dataset_originaldocs_db['region_name'].T


0    1    2   3    4    5    6    7    8
region_id          0    1    2   3    4    5    6    7    8
region_name  Unknown  CA1  CA3  DG  EC2  EC3  EC5  EC4  EC?

In [52]:
dataset_originaldocs_db['elepos'].sample(10).T


27             34             63          102           128  \
topdir       ec013.33       ec013.40       ec014.29  gor01-6-13       km01.09   
mergen  ec013.540_561  ec013.702_724  ec014.459_480   2006-6-13  km01.108_124   
unit            unit5          unit6          unit3       unit2         unit5   
fileid          20033          20040          30029       50613         70009   
e1                EC2            EC2            CA1         CA3                 
e2                EC3            EC3            CA1         CA3           CA1   
e3                EC4            EC4            CA1         CA3           CA1   
e4                EC5            EC5            CA1         CA3           CA1   
e5                CA1            CA1            CA1         CA3           CA1   
e6                CA1            CA1            CA1         CA3                 
e7                CA1            CA1            CA1         CA3                 
e8                CA1            CA1            CA1         CA3                 
e9                                              EC2         CA1                 
e10                                             EC3         CA1                 
e11                                             EC3         CA1                 
e12                                             EC5         CA1                 
e13                                                                             
e14                                                                             
e15                                                                             
e16                                                                             

                 127              5                     119            54   \
topdir       km01.08       ec012ec.16            j01_maze04       ec014.17   
mergen  km01.087_099  ec012ec.304-319  j01_maze04_M.001_004  ec014.192_204   
unit           unit5            unit6                 unit6          unit4   
fileid         70008            10016                120004          30017   
e1                                EC3                                  CA1   
e2               CA1              EC3                                  CA1   
e3               CA1              EC5                                  CA1   
e4               CA1              EC5                                  CA1   
e5               CA1                                     CA            CA1   
e6                                                       CA            CA1   
e7                                                       CA            CA1   
e8                                                       CA            CA1   
e9                                                                     EC2   
e10                                                                    EC3   
e11                                                                    EC3   
e12                                                                    EC5   
e13                                                                          
e14                                                                          
e15                                                                          
e16                                                                          

                  46   
topdir       ec013.52  
mergen  ec013.921_927  
unit            unit2  
fileid          20052  
e1                EC2  
e2                EC3  
e3                EC4  
e4                EC5  
e5                CA3  
e6                CA3  
e7                CA3  
e8                CA3  
e9                     
e10                    
e11                    
e12                    
e13                    
e14                    
e15                    
e16

In [53]:
dataset_originaldocs_db['clu_cell_counts'].sample(10).T

27        50        62        77        49        43        21  \
topdir   ec013.33  ec013.56  ec014.28  ec016.30  ec013.55  ec013.49  ec013.27   
nshanks         8         8        12        12         8         8         8   
ncells        104        96       124        93        88       118       107   

               47        18          10  
topdir   ec013.53  ec013.15  ec012ec.21  
nshanks         8         8           4  
ncells         90        99          21

In [54]:
dataset_originaldocs_db['map_cell_counts'].sample(10).T

36        48        23        76          99          16   \
topdir  ec013.42  ec013.54  ec013.29  ec016.28  f01_maze08  ec012ec.27   
ncells       106        58        93        82          31          17   

             53          104         7           68   
topdir  ec014.16  g01_maze05  ec012ec.18  ec014.n329  
ncells       132          44          25          49

In [55]:
dataset_originaldocs_db['pmMap'].sample(10).T


3873       4910       8015       8082  \
id                                    3874       4911       8016       8083   
file_ID                                 57         68        118        119   
cell_ID                                 36         60         66         18   
ele                                      2          8          7          6   
clu                                     13          5         12          9   
exciting                                 0          0          0          2   
inhibiting                               0          0          0          0   
excited                                  0          0          0          0   
inhibited                                0          0          0          3   
wide                                     1          1          0          0   
narrow                                   0          0          0          0   
IsPositive                               0          0          0          0   
eDist                            30.181736   20.92805  31.108251  13.556632   
RefracRatio                            0.0   0.020444        0.0   0.032481   
RefracViol                             0.0   0.000422        0.0   0.001641   
eDist_ind                        32.340283  26.011186        0.0        0.0   
RefracRatio_ind                        0.0        0.0        0.0        0.0   
RefracViol_ind                         0.0        0.0        0.0        0.0   
FileID                               30021      40025      60015     120004   
RegionID                                 1          1          1          0   
AnimalID                                 3          4          6         12   
FireRate                          0.076174    0.14819   0.100019   1.456594   
totalFireRate                     0.076064   0.148266   0.122313   1.459221   
Clean_G                                  1          1          0          0   
Clean_ind                                1          1          0          0   
COM_15_msec                       8.353969   8.840442        0.0        0.0   
COM_20_msec                       9.563175  10.274054        0.0        0.0   
COM_25_msec                      10.467456    11.6513        0.0        0.0   
COM_30_msec                       11.34382  12.895799        0.0        0.0   
COM_40_msec                      12.810526  15.339181        0.0        0.0   
COM_50_msec                      13.827217  17.920232        0.0        0.0   
Wide_ind_WidthR_AmpSym                 1.0        1.0        0.0        0.0   
Wide_ind_WidthR_COM                    1.0        1.0        0.0        0.0   
Wide_ind_WidthR_totalFireRate          1.0        1.0        0.0        0.0   
Wide_ind_WidthC100_AmpSym              1.0        1.0        0.0        0.0   
Narrow_ind_WidthR_AmpSym               0.0        0.0        0.0        0.0   
Narrow_ind_WidthR_COM                  0.0        0.0        0.0        0.0   
Narrow_ind_WidthR_totalFireRate        0.0        0.0        0.0        0.0   
Narrow_ind_WidthC100_AmpSym            0.0        0.0        0.0        0.0   
Wide_visual_correction                   0          0          0          0   
Narrow_visual_correction                 0          0          0          0   
Wide_final                               1          1          1          1   
Narrow_final                             0          0          0          0   
Big_Which_Region                         2          2          2          0   

                                      264        1500       5439       696   \
id                                     265       1501       5440        697   
file_ID                                 14         32         75         23   
cell_ID                                  7         23         20         36   
ele                                      3          2          5          3   
clu                                      5          8          3         17   
exciting

In [56]:
dataset_originaldocs_db['region_pm1943'].sample(10).T


3582  4172  7542  6900 237   8048  4539  2871  3833 280 
id      3583  4173  7543  6901  238  8049  4540  2872  3834  281
region   EC3   CA1   CA1   CA3  EC5  CA1c   CA1   EC3   EC5  EC3

In [57]:
dataset_originaldocs_db['region_elepos'].sample(10).T


4432 609   1737  7336  3054  2878 641   4960  8171  7489
id      4433  610  1738  7337  3055  2879  642  4961  8172  7490
region   EC2  EC5   EC5   CA1   CA3   EC3  CA1   EC3   CA1   CA1

In [58]:
dataset_originaldocs_db['clean'].sample(10).T


3864  1791  634   5630  8061  5954  1924  3450  3733  5897
id    3865  1792   635  5631  8062  5955  1925  3451  3734  5898
good     1     1     1     1     1     1     1     1     1     1

In [59]:
dataset_originaldocs_db['region'].sample(10).T


3552  2505  497   7353  1801  1282  1140  7173  2061  562 
id         3553  2506   498  7354  1802  1283  1141  7174  2062   563
region_id     1     4     5     5     5     1     5     1     1     1

In [60]:
dataset_originaldocs_db['iCell'].sample(10).T

2758  4028  8126  8545  3668  2405  2567  148   3325  6512
id           2759  4029  8127  8546  3669  2406  2568   149  3326  6513
pyramidal       0     1     1     1     0     1     0     1     1     1
interneuron     1     0     0     0     1     0     1     0     0     0
not_pyin        0     0     0     0     0     0     0     0     0     0
exciting        0     1     0     0     0     0     0     0     1     0
inhibiting      1     0     0     0     0     0     0     0     0     0

In [61]:
dataset_originaldocs_db['to_remove'].sample(10).T


13     9      14     11     10     1      5      6      4      2 
fileid  40058  40028  81104  40031  40030  20041  30103  40017  40017  20041
ele         2      2      6      2      2      6      2      3      3      7
clu         4      5     12      5      6      9      3     13      4     12

#### The [README.txt](./data/hc-3/docs/hc3-original-docs/README.txt) file

> This directory contains the original metadata files which
describe the hc3 data set.  

>These original files should NOT
be used as the main source of information about the hc3 data set
because they have not been "cleaned up" and documented.

> Specifically, they contain extra information which is not
documented, make reference to some data which is not included
in the hc3 data set, and are harder to understand
than the metadata tables provided with the data set.

> These are provided in case the extra non-documented information
is useful and to document how the metadata tables provided
with the data set were generated.

> Included in this directory are the original data files:
KenjiData.mat and ElePosition.txt file.

##### Loading ElePosition.txt

In [250]:
import pandas as pd
dataset_originaldocs_ElePosition_filepath = os.path.join(dataset_originaldocs_directory,'ElePosition.txt')
with open(dataset_originaldocs_ElePosition_filepath,'r') as f:
    dataset_originaldocs_ElePosition_content = f.read()
dataset_originaldocs_ElePosition_content = dataset_originaldocs_ElePosition_content.replace('\t\t\t','\t')
dataset_originaldocs_ElePosition_content = dataset_originaldocs_ElePosition_content.replace('\t\t','\t')
dataset_originaldocs_ElePosition_lines = dataset_originaldocs_ElePosition_content.strip().split('\n')
dataset_originaldocs_ElePosition = [l.split('\t') for l in dataset_originaldocs_ElePosition_lines]
dataset_originaldocs_ElePosition = pd.DataFrame(dataset_originaldocs_ElePosition)
dataset_originaldocs_ElePosition = dataset_originaldocs_ElePosition.fillna(value='')
dataset_originaldocs_ElePosition = dataset_originaldocs_ElePosition.replace('None','')
dataset_originaldocs_ElePosition

0                1      2       3    4    5    6    7    8  9   \
0    ec012ec.11  ec012ec.181-195  unit6  010011  EC3  EC3  EC5  EC5           
1    ec012ec.12  ec012ec.199-219  unit6  010012  EC3  EC3  EC5  EC5           
2    ec012ec.13  ec012ec.220-246  unit6  010013  EC3  EC3  EC5  EC5           
3    ec012ec.14  ec012ec.251-274  unit6  010014  EC3  EC3  EC5  EC5           
4    ec012ec.15  ec012ec.276-298  unit6  010015  EC3  EC3  EC5  EC5           
..          ...              ...    ...     ...  ...  ...  ...  ...  ... ..   
128     km01.09     km01.108_124  unit5  070009       CA1  CA1  CA1  CA1      
129    KM011203   KM011203_01_10  unit5  070101  CA1  CA1  CA1  CA1           
130    KM011204   KM011204_02_10  unit5  070102  CA1  CA1  CA1  CA1           
131    KM011205   KM011205_01_11  unit5  070103  CA1  CA1  CA1  CA1           
132    KM011206   KM011206_01_11  unit5  070104  CA1  CA1  CA1  CA1           

    10 11 12 13 14 15 16 17 18 19  
0                                  
1                                  
2                                  
3                                  
4                                  
..  .. .. .. .. .. .. .. .. .. ..  
128                                
129                                
130                                
131                                
132                                

[133 rows x 20 columns]

##### Loading KenjiData.mat

>KenjiData.mat is a MatLab file.

In [186]:
from scipy.io import loadmat

In [187]:
dataset_originaldocs_KenjiData_filepath = os.path.join(dataset_originaldocs_directory,'KenjiData.mat')

In [188]:
dataset_originaldocs_KenjiData = loadmat(dataset_originaldocs_KenjiData_filepath)

In [205]:
[print(k) for k in dataset_originaldocs_KenjiData.keys()];

__header__
__version__
__globals__
PyrIntMap
iCell
Region
Clean
Beh


##### Map, clean and icell shapes

> When loading into MatLab, you will see total　8569 cells there.

> Each row of PyrIntMap.Map, Clean, iCell{NN}, Region is the same cell.

> size(PyrIntMap.Map) =  [8569 43];

In [78]:
dataset_originaldocs_KenjiData['PyrIntMap']['Map'][0][0].shape

(8569, 43)

In [79]:
dataset_originaldocs_KenjiData['Clean'].shape

(8569, 1)

In [80]:
dataset_originaldocs_KenjiData['iCell'][0][0].shape

(8569, 1)

##### Map column names 

> What dose each column say? see PyrIntMap.content

In [191]:
print(dataset_originaldocs_KenjiData['PyrIntMap']['content'][0][0].shape)
[(f+1,field[0]) for f,field in enumerate(dataset_originaldocs_KenjiData['PyrIntMap']['content'][0][0][0])]

(1, 43)


[(1, 'file_ID'),
 (2, 'cell_ID'),
 (3, 'ele'),
 (4, 'clu'),
 (5, 'exciting'),
 (6, 'inhibiting'),
 (7, 'excited'),
 (8, 'inhibited'),
 (9, 'wide'),
 (10, 'narrow'),
 (11, 'IsPositive'),
 (12, 'eDist'),
 (13, 'RefracRatio'),
 (14, 'RefracViol'),
 (15, 'eDist_ind'),
 (16, 'RefracRatio_ind'),
 (17, 'RefracViol_ind'),
 (18, 'FileID'),
 (19, 'RegionID'),
 (20, 'AnimalID'),
 (21, 'FireRate'),
 (22, 'totalFireRate'),
 (23, 'Clean_G'),
 (24, 'Clean_ind'),
 (25, 'COM_15_msec'),
 (26, 'COM_20_msec'),
 (27, 'COM_25_msec'),
 (28, 'COM_30_msec'),
 (29, 'COM_40_msec'),
 (30, 'COM_50_msec'),
 (31, 'Wide_ind_WidthR_AmpSym'),
 (32, 'Wide_ind_WidthR-COM'),
 (33, 'Wide_ind_WidthR_totalFireRate'),
 (34, 'Wide_ind_WidthC100_AmpSym'),
 (35, 'Narrow_ind_WidthR_AmpSym'),
 (36, 'Narrow_ind_WidthR-COM'),
 (37, 'Narrow_ind_WidthR_totalFireRate'),
 (38, 'Narrow_ind_WidthC100_AmpSym'),
 (39, 'Wide_visual_correction'),
 (40, 'Narrow_visual_correction'),
 (41, 'Wide_final'),
 (42, 'Narrow_final'),
 (43, 'Big_Which_R


> PyrIntMap.Map(:,18) = File ID, See /u12/kenji/data/ElePosition.txt, 4th column.

> PyrIntMap.Map(:,3) = electrode #

> PyrIntMap,Map(:,4) = clu #

> Map(:,5): number of cells excited by this cell.

> Map(:,6): number of cells inhibited by this cell.

> 

> Stated again, contents of Map columns:

> > 1 file_ID

> file ID #

> > 2 cell_ID

> This is cell ID # in the session, starting from 1 to number of cells. Not useful at all.

> > 3 ele

> electrode #

> > 4 clu

> ID # in clu files

> > 5 exciting

> Based on cross-correlogram analysis, the cell monosynaptically excites other cells.

> > 6 inhibiting

> Based on cross-correlogram analysis, the cell monosynaptically inhibits other cells.

> > 7 excited

> Based on cross-correlogram analysis, the cell is monosynaptically excited other cells.

> > 8 inhibited

> Based on cross-correlogram analysis, the cell is monosynaptically excited other cells.

> 

> >

> >

> > 19 RegionID

> > 20 AnimalID

> 20 corresponds to PhyIntMap.WhichFile.

> > 21 FireRate

> > 22 totalFireRate

> difference bwt 21 and 22?

> 

> 21.

> MeanISI = mean(bootstrp(100,'mean',ISI));

> FireRate = SampleRate/MeanISI

> 

> ISI is interspike intervals.

> 

> This is what i did but I'm not so sure if the second method is valid.

> 

> 22. # of spikes divided by total recording length.

> 

> > 43 Big_Which_Region(EC=1,CA1=2,CA3,DG=3)

> 

> 19 and 43, correspond to PhyIntMap.whichRegion.

> 

> -------------------

> 

> 

> 

##### Regions

> Region; 1=CA1,2=CA3,3=DG,4=EC2,5=EC3,6=EC5

In [103]:
print(KenjiData['Region'].shape)
print(np.unique(KenjiData['Region']))
print([(r,np.sum(r == KenjiData['Region'])) for r in np.unique(KenjiData['Region'])])

(8569, 1)
[0 1 2 3 4 5 6 7 8]
[(0, 132), (1, 4334), (2, 1237), (3, 136), (4, 622), (5, 900), (6, 775), (7, 344), (8, 89)]


It seems a mismatch between README.txt and the Region values, the [DecidePyrIntRegion.m](./data/hc-3/docs/hc3-original-docs/DecidePyrIntRegion.m) elucidates with more values:

> Region = zeros(size(Map,1),1);
> 
> % CA1
> ind = Map(:,43)==2;
> Region(ind,:) = 1;
> 
> % CA3
> ind = Map(:,43)==3 & Map(:,19)==1;
> Region(ind,:) = 2;
> 
> % DG
> ind = Map(:,43)==3 & Map(:,19)==2 & Map(:,18)~=20051;
> Region(ind,:) = 3;
> 
> % EC2
> ind = Map(:,43)==1 & Map(:,19)==1;
> Region(ind,:) = 4;
> 
> % EC3
> ind = Map(:,43)==1 & Map(:,19)==2;
> Region(ind,:) = 5;
> 
> % EC5
> ind = Map(:,43)==1 & Map(:,19)==4;
> Region(ind,:) = 6;
> 
> % EC4
> ind = Map(:,43)==1 & Map(:,19)==3;
> Region(ind,:) = 7;
> 
> % EC?
> ind = Map(:,43)==1 & Map(:,19)>4;
> Region(ind,:) = 8;

In [104]:
KenjiData_Region_dict = { 0:'', 1:'CA1', 2:'CA3', 3:'DG', 4:'EC2', 5:'EC3', 6:'EC5', 7:'EC4', 8:'EC?' }

In [105]:
KenjiData_Region = [KenjiData_Region_dict[r[0]] for r in KenjiData['Region']]

##### Good and bad cells

> Clean: 0=bad cells

In [106]:
print(KenjiData['Clean'].shape)
print(np.unique(KenjiData['Clean']))

(8569, 1)
[0 1]


##### Cell phenotype

> iCell; iCell{1}==1,pyr cells, iCell{2}==1,interneurons

In [114]:
KenjiData['iCell'].shape
KenjiData['iCell'][0][1].shape

(8569, 1)

In [49]:
PyrIntMap = KenjiData['PyrIntMap']

In [255]:
PyrIntMap.dtype.names

('Map',
 'date',
 'fileBase',
 'whichRegion',
 'content',
 'WhichFile',
 'date2',
 'date_final',
 'data_final')

In [50]:
PyrIntMap['whichRegion']

array([[array([[array([[array(['EC2'], dtype='<U3'), array(['EC3'], dtype='<U3'),
                        array(['EC4'], dtype='<U3'), array(['EC5'], dtype='<U3'),
                        array(['EC11'], dtype='<U4'), array(['EC12'], dtype='<U4'),
                        array(['EC13'], dtype='<U4'), array(['EC14'], dtype='<U4')]],
                      dtype=object)                                                  ,
                array([[array(['CA1'], dtype='<U3'), array(['CA1c'], dtype='<U4')]],
                      dtype=object)                                                 ,
                array([[array(['CA3'], dtype='<U3'), array(['DG'], dtype='<U2')]],
                      dtype=object)                                               ]],
              dtype=object)                                                           ]],
      dtype=object)

In [256]:
{ dat_key : PyrIntMap[dat_key][0][0][0] for dat_key in PyrIntMap.dtype.names if dat_key.startswith('dat')}

{'date': '25-Dec-2008',
 'date2': '09-Apr-2010',
 'date_final': '07-Sep-2010',
 'data_final': '12-Oct-2010'}

In [259]:
print(PyrIntMap['Map'][0][0].shape)
PyrIntMap_Map = PyrIntMap['Map'][0][0]
PyrIntMap_Map

(8569, 43)


array([[  1.,   1.,   1., ...,   1.,   0.,   1.],
       [  1.,   2.,   1., ...,   0.,   1.,   1.],
       [  1.,   3.,   1., ...,   1.,   0.,   1.],
       ...,
       [133.,  47.,  11., ...,   0.,   1.,   1.],
       [133.,  48.,  11., ...,   1.,   0.,   1.],
       [133.,  49.,  11., ...,   1.,   0.,   1.]])

In [260]:
print(PyrIntMap['content'][0][0][0].shape)
PyrIntMap_content = [c[0] for c in PyrIntMap['content'][0][0][0]]
PyrIntMap_content

(43,)


['file_ID',
 'cell_ID',
 'ele',
 'clu',
 'exciting',
 'inhibiting',
 'excited',
 'inhibited',
 'wide',
 'narrow',
 'IsPositive',
 'eDist',
 'RefracRatio',
 'RefracViol',
 'eDist_ind',
 'RefracRatio_ind',
 'RefracViol_ind',
 'FileID',
 'RegionID',
 'AnimalID',
 'FireRate',
 'totalFireRate',
 'Clean_G',
 'Clean_ind',
 'COM_15_msec',
 'COM_20_msec',
 'COM_25_msec',
 'COM_30_msec',
 'COM_40_msec',
 'COM_50_msec',
 'Wide_ind_WidthR_AmpSym',
 'Wide_ind_WidthR-COM',
 'Wide_ind_WidthR_totalFireRate',
 'Wide_ind_WidthC100_AmpSym',
 'Narrow_ind_WidthR_AmpSym',
 'Narrow_ind_WidthR-COM',
 'Narrow_ind_WidthR_totalFireRate',
 'Narrow_ind_WidthC100_AmpSym',
 'Wide_visual_correction',
 'Narrow_visual_correction',
 'Wide_final',
 'Narrow_final',
 'Big_Which_Region(EC=1,CA1=2,CA3,DG=3)']

In [267]:
PyrIntMap_dataframe = pd.DataFrame(PyrIntMap_Map,columns=PyrIntMap_content)
PyrIntMap_dataframe.sample(10).T

4927          2439  \
file_ID                                   68.000000     41.000000   
cell_ID                                   77.000000     48.000000   
ele                                       11.000000      5.000000   
clu                                        2.000000      3.000000   
exciting                                   0.000000      1.000000   
inhibiting                                 0.000000      0.000000   
excited                                    0.000000      0.000000   
inhibited                                  0.000000      0.000000   
wide                                       1.000000      1.000000   
narrow                                     0.000000      0.000000   
IsPositive                                 0.000000      0.000000   
eDist                                     22.932338     23.374291   
RefracRatio                                0.286721      0.028186   
RefracViol                                 0.002105      0.000567   
eDist_ind                                 27.160555     33.165368   
RefracRatio_ind                            0.192737      0.000000   
RefracViol_ind                             0.000805      0.000000   
FileID                                 40025.000000  20046.000000   
RegionID                                   4.000000      1.000000   
AnimalID                                   4.000000      2.000000   
FireRate                                   1.367948      0.146794   
totalFireRate                              1.367505      0.192930   
Clean_G                                    1.000000      1.000000   
Clean_ind                                  1.000000      1.000000   
COM_15_msec                                9.702031      8.851744   
COM_20_msec                               12.065240     10.353462   
COM_25_msec                               14.253271     11.677466   
COM_30_msec                               16.432600     12.708418   
COM_40_msec                               20.801385     14.348504   
COM_50_msec                               25.082878     15.747024   
Wide_ind_WidthR_AmpSym                     1.000000      1.000000   
Wide_ind_WidthR-COM                        1.000000      1.000000   
Wide_ind_WidthR_totalFireRate              1.000000      1.000000   
Wide_ind_WidthC100_AmpSym                  1.000000      1.000000   
Narrow_ind_WidthR_AmpSym                   0.000000      0.000000   
Narrow_ind_WidthR-COM                      0.000000      0.000000   
Narrow_ind_WidthR_totalFireRate            0.000000      0.000000   
Narrow_ind_WidthC100_AmpSym                0.000000      0.000000   
Wide_visual_correction                     0.000000      0.000000   
Narrow_visual_correction                   0.000000      0.000000   
Wide_final                                 1.000000      1.000000   
Narrow_final                               0.000000      0.000000   
Big_Which_Region(EC=1,CA1=2,CA3,DG=3)      1.000000      2.000000   

                                               5869          4363  \
file_ID                                   85.000000     63.000000   
cell_ID                                   11.000000     11.000000   
ele                                        4.000000      1.000000   
clu                                        9.000000     12.000000   
exciting                                   0.000000      0.000000   
inhibiting                                 0.000000      0.000000   
excited                                    0.000000      0.000000   
inhibited                                  1.000000      0.000000   
wide                                       0.000000      1.000000   
narrow                                     0.000000      0.000000   
IsPositive                                 0.000000      0.000000   
eDist                                     36.947641     17.336967   
RefracRatio                                0.015333      0.184441   
RefracViol                                 0

In [268]:
print(np.min(PyrIntMap_dataframe['file_ID']))
print(np.max(PyrIntMap_dataframe['file_ID']))


1.0
133.0


In [283]:
PyrIntMap_fileBase_dict = {f+1:fb[0] for f,fb in enumerate(PyrIntMap['fileBase'][0][0][0])}
PyrIntMap_fileBase_dict

{1: 'ec012ec.181-195',
 2: 'ec012ec.199-219',
 3: 'ec012ec.220-246',
 4: 'ec012ec.251-274',
 5: 'ec012ec.276-298',
 6: 'ec012ec.304-319',
 7: 'ec012ec.325-363',
 8: 'ec012ec.365-380',
 9: 'ec012ec.384-402',
 10: 'ec012ec.405-414',
 11: 'ec012ec.416-434',
 12: 'ec012ec.436-449',
 13: 'ec012ec.452-471',
 14: 'ec012ec.473-493',
 15: 'ec012ec.495-506',
 16: 'ec012ec.513-529',
 17: 'ec012ec.532-535',
 18: 'ec012ec.546-564',
 19: 'ec013.152_157',
 20: 'ec013.194_213',
 21: 'ec013.242_258',
 22: 'ec013.370_390',
 23: 'ec013.393_418',
 24: 'ec013.421_447',
 25: 'ec013.451_470',
 26: 'ec013.480_509',
 27: 'ec013.512_537',
 28: 'ec013.540_561',
 29: 'ec013.564_585',
 30: 'ec013.588_604',
 31: 'ec013.608_632',
 32: 'ec013.634_653',
 33: 'ec013.656_674',
 34: 'ec013.677_699',
 35: 'ec013.702_724',
 36: 'ec013.733_746',
 37: 'ec013.752_769',
 38: 'ec013.771_777',
 39: 'ec013.779_791',
 40: 'ec013.793_814',
 41: 'ec013.816_835',
 42: 'ec013.839_851',
 43: 'ec013.853_866',
 44: 'ec013.868_893',
 45: 

In [285]:
PyrIntMap_dataframe['file_base'] = PyrIntMap_dataframe['file_ID'].astype(int).map(lambda f: PyrIntMap_fileBase_dict[f])
PyrIntMap_dataframe['file_base']

0       ec012ec.181-195
1       ec012ec.181-195
2       ec012ec.181-195
3       ec012ec.181-195
4       ec012ec.181-195
             ...       
8564         nlx_070329
8565         nlx_070329
8566         nlx_070329
8567         nlx_070329
8568         nlx_070329
Name: file_base, Length: 8569, dtype: object

In [205]:
(
    'fileBase',
    'whichRegion',
    'WhichFile',
)

('fileBase', 'whichRegion', 'WhichFile')

In [154]:
[(f+1,fb[0]) for f,fb in enumerate(PyrIntMap['fileBase'][0][0][0])]

[(1, 'ec012ec.181-195'),
 (2, 'ec012ec.199-219'),
 (3, 'ec012ec.220-246'),
 (4, 'ec012ec.251-274'),
 (5, 'ec012ec.276-298'),
 (6, 'ec012ec.304-319'),
 (7, 'ec012ec.325-363'),
 (8, 'ec012ec.365-380'),
 (9, 'ec012ec.384-402'),
 (10, 'ec012ec.405-414'),
 (11, 'ec012ec.416-434'),
 (12, 'ec012ec.436-449'),
 (13, 'ec012ec.452-471'),
 (14, 'ec012ec.473-493'),
 (15, 'ec012ec.495-506'),
 (16, 'ec012ec.513-529'),
 (17, 'ec012ec.532-535'),
 (18, 'ec012ec.546-564'),
 (19, 'ec013.152_157'),
 (20, 'ec013.194_213'),
 (21, 'ec013.242_258'),
 (22, 'ec013.370_390'),
 (23, 'ec013.393_418'),
 (24, 'ec013.421_447'),
 (25, 'ec013.451_470'),
 (26, 'ec013.480_509'),
 (27, 'ec013.512_537'),
 (28, 'ec013.540_561'),
 (29, 'ec013.564_585'),
 (30, 'ec013.588_604'),
 (31, 'ec013.608_632'),
 (32, 'ec013.634_653'),
 (33, 'ec013.656_674'),
 (34, 'ec013.677_699'),
 (35, 'ec013.702_724'),
 (36, 'ec013.733_746'),
 (37, 'ec013.752_769'),
 (38, 'ec013.771_777'),
 (39, 'ec013.779_791'),
 (40, 'ec013.793_814'),
 (41, 'ec013.

In [171]:
print(PyrIntMap['whichRegion'][0][0][0].shape)
[list(r[0]) for r in PyrIntMap['whichRegion'][0][0][0]]

(3,)


[[array(['EC2'], dtype='<U3'),
  array(['EC3'], dtype='<U3'),
  array(['EC4'], dtype='<U3'),
  array(['EC5'], dtype='<U3'),
  array(['EC11'], dtype='<U4'),
  array(['EC12'], dtype='<U4'),
  array(['EC13'], dtype='<U4'),
  array(['EC14'], dtype='<U4')],
 [array(['CA1'], dtype='<U3'), array(['CA1c'], dtype='<U4')],
 [array(['CA3'], dtype='<U3'), array(['DG'], dtype='<U2')]]

In [176]:
print(PyrIntMap['Map'][0][0].shape)
PyrIntMap['Map'][0][0]

(8569, 43)


array([[  1.,   1.,   1., ...,   1.,   0.,   1.],
       [  1.,   2.,   1., ...,   0.,   1.,   1.],
       [  1.,   3.,   1., ...,   1.,   0.,   1.],
       ...,
       [133.,  47.,  11., ...,   0.,   1.,   1.],
       [133.,  48.,  11., ...,   1.,   0.,   1.],
       [133.,  49.,  11., ...,   1.,   0.,   1.]])

size(PyrIntMap.Map) =  [8569 43];

What dose each column say? see PyrIntMap.content

PyrIntMap.Map(:,18) = File ID, See /u12/kenji/data/ElePosition.txt, 4th column.

PyrIntMap.Map(:,3) = electrode #

PyrIntMap,Map(:,4) = clu #

Map(:,5): number of cells excited by this cell.

Map(:,6): number of cells inhibited by this cell.



Stated again, contents of Map columns:

> 1 file_ID

file ID #

> 2 cell_ID

This is cell ID # in the session, starting from 1 to number of cells. Not useful at all.

> 3 ele

electrode #

> 4 clu

ID # in clu files

> 5 exciting

Based on cross-correlogram analysis, the cell monosynaptically excites other cells.

> 6 inhibiting

Based on cross-correlogram analysis, the cell monosynaptically inhibits other cells.

> 7 excited

Based on cross-correlogram analysis, the cell is monosynaptically excited other cells.

> 8 inhibited

Based on cross-correlogram analysis, the cell is monosynaptically excited other cells.



>

>

> 19 RegionID

> 20 AnimalID

20 corresponds to PhyIntMap.WhichFile.

> 21 FireRate

> 22 totalFireRate

difference bwt 21 and 22?



21.

MeanISI = mean(bootstrp(100,'mean',ISI));

FireRate = SampleRate/MeanISI

ISI is interspike intervals.



This is what i did but I'm not so sure if the second method is valid.



22. of spikes divided by total recording length.



> 43 Big_Which_Region(EC=1,CA1=2,CA3,DG=3)



19 and 43, correspond to PhyIntMap.whichRegion.

In [122]:
KenjiData_matfile['Clean'].shape

(8569, 1)

In [123]:
print(KenjiData_matfile['iCell'].shape)
print(KenjiData_matfile['iCell'][0][0].shape)
print(KenjiData_matfile['iCell'][0][1].shape)
print(KenjiData_matfile['iCell'][0][2].shape)
print(KenjiData_matfile['iCell'][0][3].shape)
print(KenjiData_matfile['iCell'][0][4].shape)

(1, 5)
(8569, 1)
(8569, 1)
(8569, 1)
(8569, 1)
(8569, 1)


In [124]:
KenjiData_matfile['Region'].shape

(8569, 1)

In [125]:
KenjiData_matfile['Beh'].shape

(1692, 8)

In [126]:
{k : v for k,v in KenjiData_matfile.items()};

In [127]:
[x for x in pd.DataFrame(KenjiData_matfile['Beh']).iloc[:,7].values if x != '']

/tmp/ipykernel_8198/3036829709.py:1: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  [x for x in pd.DataFrame(KenjiData_matfile['Beh']).iloc[:,7].values if x != '']


[array(['1'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['video'], dtype='<U5'),
 array(['1'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['1'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['2'], dtype='<U1'),
 array(['1'], dtype='<U1')]